readme.txt
=== EMA4WP: ZozoEMA Top Bar ===
Contributors: Ibericode, DvanKooten, hchouhan, lapzor
Donate link: https://zozo.vn/ung-dung-email-marketing
Tags: zozoema, form, newsletter, ema4wp, email, opt-in, subscribe, call to action
Requires at least: 1.0.1
Tested up to: 1.0.1
Stable tag: 1.0.1
License: GPLv2 or later
License URI: https://www.gnu.org/licenses/gpl-2.0.html
Requires PHP: 5.3

Adds a ZozoEMA opt-in form to the top or bottom of your WordPress site.

== Description ==

Adds a beautiful, customizable sign-up bar to the top of your WordPress site. This bar is guaranteed to get the attention of your visitor and
increase your ZozoEMA subscribers.

> This plugin is an add-on for the [EMA4WP: ZozoEMA for WordPress plugin](https://wordpress.org/plugins/zozoema-for-wp/).

= ZozoEMA Sign-Up Bar, at a glance.. =

ZozoEMA Top Bar adds a simple yet beautiful & customizable opt-in bar to the top or bottom of your WordPress site.

Using this bar, people can subscribe to a ZozoEMA list of your choice.

- Guaranteed to boost conversions.
- Unobtrusive, visitors can easily dismiss the bar.
- Easy to install & configure, just select a ZozoEMA list and you're good to.
- Customizable, you can edit the bar text and colors from the plugin settings.
- The bar can be at the top or bottom of the visitor's screen
- Lightweight, the plugin consists of just a single 4kb JavaScript file.

= Development of ZozoEMA Top Bar =

Bug reports (and Pull Requests) for [ZozoEMA Top Bar are welcomed on GitHub](https://github.com/lethi2k/ema-top-bar). Please note that GitHub is _not_ a support forum.

**More information**

- [EMA4WP: ZozoEMA for WordPress](https://wordpress.org/plugins/zozoema-for-wp/)
- Developers; follow or contribute to the [ZozoEMA Top Bar plugin on GitHub]https://github.com/lethi2k/ema-top-bar)

== Installation ==

= ZozoEMA for WordPress =

Since this plugin depends on the [ZozoEMA for WordPress plugin](https://wordpress.org/plugins/zozoema-for-wp/), you will need to install that first.

= Installing ZozoEMA Top Bar =

1. In your WordPress admin panel, go to *Plugins > New Plugin*, search for **ZozoEMA Top Bar** and click "*Install now*"
1. Alternatively, download the plugin and upload the contents of `zozoema-top-bar.zip` to your plugins directory, which usually is `/wp-content/plugins/`.
1. Activate the plugin
1. Set [your ZozoEMA API key](https://appv4.zozo.vn/account/api) in **ZozoEMA for WP > ZozoEMA Settings**.
1. Select a ZozoEMA list to subscribe to in **ZozoEMA for WP > Top Bar**.
1. _(Optional)_ Customize the look & position of your opt-in bar.

== Frequently Asked Questions ==

= How to disable the bar on certain pages? =

For now, you will have to use a filter to disable the bar on certain pages. The following example only loads the Top Bar on your blog post pages.

`
add_filter( 'ematb_show_bar', function( $show ) {
	return is_single();
} );
`

Another example, this only loads the bar on your "contact" page.

`
add_filter( 'ematb_show_bar', function( $show ) {
	return is_page('contact');
} );
`

Have a look at the [Conditional Tags](https://appv4.zozo.vn/frontend/docs/api/v1) page for all accepted functions.

= How to add a name field to the bar? =

You can use the following code snippet to show a "NAME" field in your bar.

`
add_action( 'ematb_before_submit_button', function() {
    echo '<input type="text" name="NAME" placeholder="Your name" />';
});

add_filter( 'ematb_subscriber_data', function( $subscriber ) {
    if( ! empty( $_POST['NAME'] ) ) {
        $subscriber->merge_fields['NAME'] = sanitize_text_field( $_POST['NAME'] );
    }

    return $subscriber;
});
`

**KB:** [Add name field to ZozoEMA Top Bar](https://appv4.zozo.vn/frontend/docs/api/v1)

= How to hide the bar on small screens? =

Adding the following CSS to your site should hide the bar on all screens smaller than 600px. The [Simple Custom CSS](https://wordpress.org/plugins/simple-custom-css/) plugin is great for adding custom CSS.

`
@media( max-width: 600px ) {
	&#35;zozoema-top-bar { display: none !important; }
}
`

= I think I found a bug. What now? =

Please report it on [GitHub issues](https://github.com/lethi2k/ema-top-bar) if it's not in the list of known issues.

= I have another question =

Please open a topic on the [WordPress.org plugin support forums](https://github.com/lethi2k/ema-top-bar).


== Screenshots ==

1. The ZozoEMA Top Bar in action on the [ZozoEMA for WordPress site](https://appv4.zozo.vn/lists).
2. The settings page of the ZozoEMA Top Bar plugin.

== Changelog ==


#### 1.0.1 - Mar 9, 2020

- Add setting to disable bar (stop loading it altogether) after it is used.
- Increase default cookie lifetime to 1 year.



CHANGELOG.md
Changelog
==========

#### 1.5.2 - Mar 9, 2020

- Add setting to disable bar (stop loading it altogether) after it is used.
- Increase default cookie lifetime to 1 year.


#### 1.5.1 - Jan 21, 2020

- Fade response element using CSS animations for better performance.
- Various minor performance improvements.


#### 1.5.0 - Oct 7, 2019

Compatibility with [ZozoEMA for WordPress](https://wordpress.org/plugins/zozoema-for-wp/) version 1.0.1.


#### 1.4.1 - Sep 11, 2019

**Changes**

- Change name to EMA4WP: ZozoEMA Top Bar.


#### 1.4.0 - Sep 4, 2019

**Improvements**

- Add (advanced) setting to quickly disable the top bar on certain pages.


#### 1.3.2 - Aug 8, 2018

**Fixes**

- Required fields notice on selected list was not showing because of invalid list property.

**Improvements**

- Prefix internal CSS classes for improved compatibility with other themes or plugins applying global admin styles.


#### 1.3.1 - May 29, 2018

**Improvements**

- 30% reduction in script file size because of removed JS dependency.
- Stop setting unused cookie when Top Bar form is used to subscribe.
- Add ematb_after_submit_button action hook.
- Improve animation performance.


#### 1.3 - November 1, 2017

**Improvements**

- Form now submits over AJAX, no longer reloading the entire page.
- Added `for` attribute to label elements, thanks [gabriel-kaam](https://github.com/gabriel-kaam).
- Added `ematb_replace_interests` filter hook.

#### 1.2.16 - January 19, 2017

Various minor code improvements.


#### 1.2.15 - September 8, 2016

**Improvements**

- Improved responsiveness when bar has additional input fields.
- Add `required` attribute to email input.


#### 1.2.14 - August 29, 2016

**Fixes**

- Top padding for small screens with admin bar.

**Improvements**

- Better bar responsiveness when window dimensions change on the fly (eg resizing a window or changing device orientation mode). (Thanks [tech4him1](https://github.com/tech4him1)!)


#### 1.2.13 - August 2, 2016

**Fixes**

- Error in animating body padding back to its original value.


#### 1.2.12 - July 21, 2016

**Fixes**

- Bar would crash when clicking toggle icon during bar animation.

**Improvements**

- Function scope generated JavaScript file to prevent Browserify clashes with other loaded scripts.
- Make sure script works even though it's loaded in the head section.
- Preparations for upcoming ZozoEMA for WordPress v4.0 release.

**Additions**

- Added Spanish language files, thanks to [Ángel Guzmán Maeso](http://shakaran.net/)
- Added `ematb_data` filter, to filter form data before it is processed.

**Deprecated**

- Deprecated `ematb_merge_vars` filter.


#### 1.2.11 - July 8, 2016

**Improvements**

- Completely removed optional jQuery dependency. The plugin now uses JavaScript animations, resulting in a much smoother experience.

#### 1.2.10 - April 12, 2016

**Fixes**

- Closed bar would still overlap underlying elements (like fixed top menu's).


#### 1.2.9 - March 16, 2016

**Fixes**

Top Bar was invisible on some themes because of `z-index` being too low.


#### 1.2.8 - March 15, 2016

**Improvements**

- Make sure top bar doesn't appear on top of WP admin bar.
- Hardened CSS styles for improved theme compatability.


#### 1.2.7 - January 26, 2016

**Improvements**

- Miscellaneous code improvements

**Additions**

- Add support for new [debug log](https://www.ema4wp.com/kb/how-to-enable-log-debugging/) in ZozoEMA for WordPress 3.1


#### 1.2.6 - January 4, 2016

 **Additions**

 - Option to "update existing subscribers" in ZozoEMA, which is useful if you have added fields.

 **Improvements**

 - Toggle icon now has a background color, for increased visibility.
 - Toggle icon now stacks above or below bar on small screens.

#### 1.2.5 - December 10, 2015

The plugin now requires [ZozoEMA for WordPress](https://wordpress.org/plugins/zozoema-for-wp/) version 1.0.1 or higher.

**Fixes**

- Fixed column alignment in Appearance tab, thanks [Chantal Coolsma](https://github.com/chantalcoolsma)!

**Improvements**

- Improved admin notice when dependencies are not installed.


#### 1.2.4 - November 22, 2015

- Compatibility for [the upcoming ZozoEMA for WordPress 3.0 release](https://www.ema4wp.com/blog/breaking-backwards-compatibility-in-version-3-0/) tomorrow.
- Added `ematb_subscribed` filter.

#### 1.2.3 - November 13, 2015

**Improvements**

- Minor refactoring in the way the plugin is bootstrapped.

#### 1.2.2 - September 10, 2015

**Fixes**

- Honeypot field being auto-completed in some browsers.
- Honeypot field was accessible by pressing "tab" key.
- Hardened security for cookie that tracks sign-up attempts.

#### 1.2.1 - September 8, 2015

**Fixes**

- Response message was not showing for some themes.

**Improvements**

- Better mobile responsiveness


#### 1.2 - September 3, 2015

**Improvements**

- The bar will now auto-dismiss after every successful sign-up.
- Placeholders will now work in Internet Explorer 7, 8 & 9 as well.

**Additions**

- Added options for double opt-in and sending ZozoEMA's "welcome email".
- Added `ematb_before_label` action allowing you to add HTML before the label-element.
- Added `ematb_before_email_field` action allowing you to add HTML before the email field.
- Added `ematb_before_submit_button` action allowing you to add HTML before the submit button.
- Added `ematb_form_action` filter allowing you to set a custom form action.

#### 1.1.3 - June 23, 2015

**Fixes**

- Fixes fatal error when visiting settings page on some servers

#### 1.1.2 - June 18, 2015

**Improvements**

- Fixes height of response message
- CSS improvements for compatibility with various popular themes

#### 1.1.1 - June 12, 2015

**Fixes**

- Fixes unclickable admin bar (or fixed navigation menu's).

**Improvements**

- Various improvements to bar CSS so it can be easily overridden.
- Fix vertical alignment of toggle icon.

#### 1.1 - June 10, 2015

**Improvements**


src/Bar.php
<?php
/*
ZozoEMA Top Bar
Copyright (C) 2015, Danny van Kooten, hi@dannyvankooten.com

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 1 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
namespace ZozoEMA\TopBar;

use Exception;
use EMA4WP_ZozoEMA;
use EMA4WP_Debug_Log;
use EMA4WP_ZozoEMA_Subscriber;
use EMA4WP_List_Data_Mapper;

class Bar {
	/**
	 * @var bool
	 */
	private $success = false;

	/**
	 * @var string
	 */
	private $error_type = '';

	/**
	 * @var bool
	 */
	private $submitted = false;

	/**
	 * Add the hooks
	 */
	public function add_hooks() {
		add_action( 'wp', array( $this, 'init' ) );
	}

	/**
	 * Add template related hooks
	 */
	public function add_template_hooks() {
		add_action( 'wp_enqueue_scripts', array( $this, 'load_assets' ) );
		add_action( 'wp_head', array( $this, 'output_css'), 90 );
		add_action( 'wp_footer', array( $this, 'output_html' ), 1 );
	}

	/**
	 * @return bool
	 */
	public function init() {

		if( ! $this->should_show_bar() ) {
			return false;
		}

		$this->add_template_hooks();
		$this->listen();
		return true;
	}

	/**
	 * Should the bar be shown?
	 *
	 * @return bool
	 */
	public function should_show_bar() {
	    $options = ematb_get_options();

		// don't show if bar is disabled
		if( ! $options['enabled'] ) {
			return false;
		}

		$show_bar = true;

		if (!empty($options['disable_on_pages'])) {
            $disable_on_pages = explode(',', $options['disable_on_pages']);
            $disable_on_pages = array_map('trim', $disable_on_pages);
            $show_bar = !is_page($disable_on_pages);
		}

		if ($options['disable_after_use'] && isset($_COOKIE['ematb_bar_hidden']) && $_COOKIE['ematb_bar_hidden'] === 'used') {
		    $show_bar = false;
        }

		/**
		 * @deprecated 1.1
		 * @use `ematb_show_bar`
		 */
        $show_bar = apply_filters( 'mctp_show_bar', $show_bar );


		/**
		 * @filter `ematb_show_bar`
		 * @expects boolean
		 *
		 * Set to true if the bar should be loaded for this request, false if not.
		 */
		return apply_filters( 'ematb_show_bar', $show_bar );
	}

	/**
	 * Listens for actions to take
	 */
	public function listen() {

		if( ! isset( $_POST['_ematb'] ) || $_POST['_ematb'] != 1 ) {
            return;
		}

		$options = ematb_get_options();
        $this->success = $this->process();

        if( ! empty( $_SERVER['HTTP_X_REQUESTED_WITH'] ) && $_SERVER['HTTP_X_REQUESTED_WITH'] === 'XMLHttpRequest' ) {
            $data = array(
                'message' => $this->get_response_message(),
                'success' => $this->success,
                'redirect_url' => $this->success ? $options['redirect']: '',
            );

            wp_send_json( $data );
            exit;
        }

        if( $this->success ) {
            // should we redirect
            $redirect_url = $options['redirect'];
            if( ! empty( $redirect_url ) ) {
                wp_redirect( $redirect_url );
                exit;
            }
        }

	}

	/**
	 * Process a form submission
	 * @return boolean
	 */
	private function process() {
        $options = ematb_get_options();
		$this->submitted = true;
		$log = $this->get_log();

		/** @var EMA4WP_ZozoEMA_Subscriber $subscriber_data */
		$subscriber = null;
		$result = false;

		if( ! $this->validate() ) {

			if( $log ) {
				$log->info( sprintf( 'Top Bar > Submitted with errors: %s', $this->error_type ) );
			}

			return false;
		}

		/**
		 * Filters the list to which ZozoEMA Top Bar subscribes.
		 *
		 * @param string $list_id
		 */
		$zozoema_list_id = apply_filters( 'ematb_zozoema_list', $options['list'] );

		// check if a ZozoEMA list was given
//		if( empty( $zozoema_list_id ) ) {
//			$this->error_type = 'error';
//
//			if( $log ) {
//				$log->warning( 'Top Bar > No ZozoEMA lists were selected' );
//			}
//
//			return false;
//		}

		$email_address = sanitize_text_field( $_POST['email'] );
		$data = array(
			'EMAIL' => $email_address,
		);

		/**
		 * Filters the data received by ZozoEMA Top Bar, before it is further processed.
		 *
		 * @param $data
		 */
		$data = apply_filters( 'ematb_data', $data );

		/** @ignore */
		$data = apply_filters( 'ematb_merge_vars', $data );
		$email_type = apply_filters( 'ematb_email_type', 'html' );

        $replace_interests = true;

        /**
         * Filters whether interests should be replaced or appended to.
         *
         * @param bool $replace_interests
         */
		$replace_interests = apply_filters( 'ematb_replace_interests', $replace_interests );

		$zozoema = new EMA4WP_ZozoEMA();
		if( class_exists( 'EMA4WP_ZozoEMA_Subscriber' ) ) {

			$mapper = new EMA4WP_List_Data_Mapper( $data, array( $zozoema_list_id ) );
			$map = $mapper->map();

			foreach( $map as $list_id => $subscriber ) {
				$subscriber->email_type = $email_type;
				$subscriber->status = $options['double_optin'] ? 'pending' : 'subscribed';

				// TODO: Add IP address.

				/** @ignore (documented elsewhere) */
				$subscriber = apply_filters( 'ema4wp_subscriber_data', $subscriber );

				/**
				 * Filter subscriber data before it is sent to ZozoEMA. Runs only for ZozoEMA Top Bar requests.
				 *
				 * @param EMA4WP_ZozoEMA_Subscriber
				 */
				$subscriber = apply_filters( 'ematb_subscriber_data', $subscriber );

				$result = $zozoema->list_subscribe( $zozoema_list_id, $subscriber->email_address, $subscriber->to_array(), $options['update_existing'], $replace_interests );
				$result = is_object( $result ) && ! empty( $result->id );
			}

		} else {
			// for BC with ZozoEMA for WordPress 3.x, override $zozoema var
			$zozoema = ema4wp_get_api();
			unset( $data['EMAIL'] );
			$result = $zozoema->subscribe( $zozoema_list_id, $email_address, $data, $email_type, $options['double_optin'], $options['update_existing'], $replace_interests, $options['send_welcome']);
		}

		// return true if success..
		if( $result ) {

			/**
			 * Fires for every successful sign-up using Top Bar.
			 *
			 * @param string $zozoema_list_id
			 * @param string $email
			 * @param array $data
			 */
			do_action( 'ematb_subscribed', $zozoema_list_id, $email_address, $data );

			// log sign-up attempt
			if( $log ) {
				$log->info( sprintf( 'Top Bar > Successfully subscribed %s', $email_address ) );
			}

			return true;
		}

		// An API error occured... Oh noes!
		if( $zozoema->get_error_code() === 214 ) {
			$this->error_type = 'already_subscribed';

			if( $log ) {
				$log->warning( sprintf( 'Top Bar > %s is already subscribed to the selected list(s)', $email_address ) );
			}
		} else {
			$this->error_type = 'error';

			if( $log ) {
				$log->error( sprintf( 'Top Bar > ZozoEMA API error: %s', $zozoema->get_error_message() ) );
			}
		}

		return false;
	}

	/**
	 * Validate the form submission
	 * @return boolean
	 */
	private function validate() {

		// make sure `email_confirm` field is given but not filled (honeypot)
		if( ! isset( $_POST['email_confirm'] ) || '' !== $_POST['email_confirm'] ) {
			$this->error_type = 'spam';
			return false;
		}

		// make sure `_ematb_timestamp` is at least 1.5 seconds ago
		if( empty( $_POST['_ematb_timestamp'] ) || time() < ( intval( $_POST['_ematb_timestamp'] ) + 1.5 ) ) {
			$this->error_type = 'spam';
			return false;
		}

		// don't work for users without JavaScript (since bar is hidden anyway, must be a bot)
		if( isset( $_POST['_ematb_no_js'] ) ) {
			$this->error_type = 'spam';
			return false;
		}

		// simple user agent check
		$user_agent = ( isset( $_SERVER['HTTP_USER_AGENT'] ) ) ? substr( $_SERVER['HTTP_USER_AGENT'], 0, 254 ) : '';
		if( strlen( $user_agent ) < 2 ) {
			$this->error_type = 'spam';
			return false;
		}

		// check if email is given and valid
		if( empty( $_POST['email'] ) || ! is_string( $_POST['email'] ) || ! is_email( $_POST['email'] ) ) {
			$this->error_type = 'invalid_email';
			return false;
		}

		return apply_filters( 'ematb_validate', true );
	}

	/**
	 * Loads the required scripts & styles
	 */
	public function load_assets() {
        $options = ematb_get_options();
		$min = ( defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ) ? '' : '.min';
		wp_enqueue_style( 'zozoema-top-bar', $this->asset_url( "/css/bar{$min}.css" ), array(), ZOZOEMA_TOP_BAR_VERSION );
		wp_enqueue_script( 'zozoema-top-bar', $this->asset_url( "/js/script{$min}.js" ), array(), ZOZOEMA_TOP_BAR_VERSION, true );

		$bottom = $options['position'] === 'bottom';

		$data = array(
			'cookieLength' => $options['cookie_length'],
			'icons' => array(
				'hide' => ( $bottom ) ? '&#x25BC;' : '&#x25B2;',
				'show' =>  ( $bottom ) ? '&#x25B2;' : '&#x25BC;'
			),
			'position' => $options['position'],
			'state' => array(
                'submitted' => $this->submitted,
                'success' => $this->success,
            ),
		);

		/**
		 * @filter `ematb_bar_config`
		 * @expects array
		 *
		 * Can be used to filter the following values
		 *  - cookieLength: The length of the cookie
		 *  - icons: Array with `hide` and `show` keys. Holds the hide/show icon strings.
		 */
		$data = apply_filters( 'ematb_bar_config', $data );

		wp_localize_script( 'zozoema-top-bar', 'ematb', $data );
	}

	/**
	 * @return string
	 */
	private function get_css_class() {
        $options = ematb_get_options();
		$classes = array( 'ematb' );

		// add class when bar is sticky
		if ($options['sticky']) {
			$classes[] = 'ematb-sticky';
		}

		// add unique css class for position (bottom|top)
		if ( $options['position']) {
			$classes[] = sprintf( 'ematb-position-%s', $options['position']);
		}

		// add class describing size of the bar
		$classes[] = "ematb-{$options['size']}";

		return join( ' ', $classes );
	}

	/**
	 * Output the CSS settings for the bar
	 */
	public function output_css() {
        $options = ematb_get_options();
		$bar_color = $options[ 'color_bar' ];
		$button_color = $options['color_button'];
		$text_color = $options[ 'color_text' ];
		$button_text_color = $options[ 'color_button_text' ];

		echo '<style type="text/css">' . PHP_EOL;

		if( ! empty( $bar_color) ) {
			echo ".ematb-bar, .ematb-response, .ematb-close { background: {$bar_color} !important; }" . PHP_EOL;
		}

		if( ! empty( $text_color ) ) {
			echo ".ematb-bar, .ematb-label, .ematb-close { color: {$text_color} !important; }" . PHP_EOL;
		}

		if( ! empty( $button_color ) ) {
			echo ".ematb-button { background: {$button_color} !important; border-color: {$button_color} !important; }" . PHP_EOL;
			echo ".ematb-email:focus { outline-color: {$button_color} !important; }" . PHP_EOL;
		}

		if( ! empty( $button_text_color ) ) {
			echo ".ematb-button { color: {$button_text_color} !important; }" . PHP_EOL;
		}

		echo '</style>';
	}


	/**
	 * Output the HTML for the opt-in bar
	 */
	public function output_html() {

		$form_action = apply_filters( 'ematb_form_action', null );
        $options = ematb_get_options();
		?>
		<div id="zozoema-top-bar" class="<?php echo $this->get_css_class(); ?>">
			<!-- ZozoEMA Top Bar v<?php echo ZOZOEMA_TOP_BAR_VERSION; ?> - https://wordpress.org/plugins/zozoema-top-bar/ -->
			<div class="ematb-bar" style="display: none">
				<form method="post" <?php if( is_string( $form_action ) ) { printf( 'action="%s"', esc_attr( $form_action ) ); } ?>>
					<?php do_action( 'ematb_before_label' ); ?>
					<label class="ematb-label" for="zozoema-top-bar__email"><?php echo $options[ 'text_bar' ]; ?></label>
					<?php do_action( 'ematb_before_email_field' ); ?>
					<input type="email" name="email" placeholder="<?php echo esc_attr( $options[ 'text_email_placeholder' ] ); ?>" class="ematb-email" required id="zozoema-top-bar__email" />
					<input type="text"  name="email_confirm" placeholder="Confirm your email" value="" autocomplete="off" tabindex="-1" class="ematb-email-confirm" />
					<?php do_action( 'ematb_before_submit_button' ); ?>
					<input type="submit" value="<?php echo esc_attr( $options['text_button'] ); ?>" class="ematb-button" />
					<?php do_action( 'ematb_after_submit_button' ); ?>
					<input type="hidden" name="_ematb" value="1" />
					<input type="hidden" name="_ematb_no_js" value="1" />
					<input type="hidden" name="_ematb_timestamp" value="<?php echo time(); ?>" />
				</form>
				<?php echo $this->get_response_message_html(); ?>
			</div>
			<!-- / ZozoEMA Top Bar -->
		</div>
		<?php
	}

	/**
	 * @return string
	 */
	protected function get_response_message() {

		if( ! $this->submitted ) {
			return '';
		}

        $options = ematb_get_options();

		if( $this->success ) {
			$message = $options['text_subscribed'];
		} else if( $this->error_type === 'already_subscribed' ) {
			$message = $options['text_already_subscribed'];
		} else if( $this->error_type === 'invalid_email' ) {
			$message = $options['text_invalid_email'];
		} else {
			$message = $options['text_error'];
		}

		return $message;
	}

	protected function get_response_message_html() {
        $message = $this->get_response_message();
        if( empty( $message ) ) {
            return '';
        }

        return sprintf( '<div class="ematb-response"><label class="ematb-response-label">%s</label></div>', $message );
    }



	/**
	 * @param $url
	 *
	 * @return string
	 */
	protected function asset_url( $url ) {
		return plugins_url( '/assets' . $url, ZOZOEMA_TOP_BAR_FILE );
	}

	/**
	 * Returns the debug logger or null, if ZozoEMA for WordPress 3.1 is not installed.
	 *
	 * @return EMA4WP_Debug_Log|null
	 */
	protected function get_log() {

		try {
			$log = ema4wp('log');
		} catch( Exception $e ) {
			return null;
		}

		return $log;
	}

}
